quickwit/quickwit-search/src/search_permits.rs

```diff
@@ -0,0 +1,213 @@
+    pub fn get_one_permit(&self) -> oneshot::Receiver<SearchPermit> {
+        let mut permits_lock = self.inner.lock().unwrap();
+        permits_lock.get_permit(&self.inner)
+    }
+
+    /// Returns a list of future permits in the form of a Receiver channel.
+    ///
+    /// The permits returned are guaranteed to be resolved in order.
+    /// In addition, the permits are guaranteed to be resolved before permits returned by
+    /// subsequent calls to this function (or get_permit).
+    pub fn get_permits_futures(&self, num_permits: usize) -> Vec<oneshot::Receiver<SearchPermit>> {
+        let mut permits_lock = self.inner.lock().unwrap();
+        permits_lock.get_permits(num_permits, &self.inner)
+    }
+}
+
+struct InnerSearchPermits {
+    num_permits_available: usize,
+    permits_requests: VecDeque<oneshot::Sender<SearchPermit>>,
+}
+
+impl InnerSearchPermits {
+    fn get_permits(
+        &mut self,
+        num_permits: usize,
+        inner: &Arc<Mutex<InnerSearchPermits>>,
+    ) -> Vec<oneshot::Receiver<SearchPermit>> {
+        let mut permits = Vec::with_capacity(num_permits);
+        for _ in 0..num_permits {
+            let (tx, rx) = oneshot::channel();
+            self.permits_requests.push_back(tx);
+            permits.push(rx);
+        }
+        self.assign_available_permits(inner);
+        permits
+    }
+
+    fn get_permit(
+        &mut self,
+        inner: &Arc<Mutex<InnerSearchPermits>>,
+    ) -> oneshot::Receiver<SearchPermit> {
+        let (tx, rx) = oneshot::channel();
+        self.permits_requests.push_back(tx);
+        self.assign_available_permits(inner);
+        rx
+    }
+
+    fn recycle_permit(&mut self, inner: &Arc<Mutex<InnerSearchPermits>>) {
+        self.num_permits_available += 1;
+        self.assign_available_permits(inner);
+    }
+
+    fn assign_available_permits(&mut self, inner: &Arc<Mutex<InnerSearchPermits>>) {
+        while self.num_permits_available > 0 {
+            let Some(sender) = self.permits_requests.pop_front() else {
+                break;
+            };
+            let send_res = sender.send(SearchPermit {
+                _ongoing_gauge_guard: GaugeGuard::from_gauge(
+                    &crate::SEARCH_METRICS.leaf_search_single_split_tasks_ongoing,
+                ),
+                inner_opt: Some(Arc::downgrade(inner)),
+            });
+            match send_res {
+                Ok(()) => {
+                    self.num_permits_available -= 1;
+                }
+                Err(mut search_permit) => {
+                    search_permit.disable_drop();
+                    drop(search_permit);
+                }
+            }
+        }
+        crate::SEARCH_METRICS
+            .leaf_search_single_split_tasks_pending
+            .set(self.permits_requests.len() as i64);
+    }
+}
+
+pub struct SearchPermit {
+    _ongoing_gauge_guard: GaugeGuard<'static>,
+    inner_opt: Option<Weak<Mutex<InnerSearchPermits>>>,
```

Comment:
I don't believe there is need for Weak here: `InnerSearchPermits` doesn't hold any `SearchPermit` (a oneshot sender doesn't hold the thing to transmit, the oneshot receiver do)

i find this type, `Option<Weak<Mutex<InnerSearchPermits>>>` rather confusing, in part because SearchPermit**s** isn't the right name imo, in part because the `Option<>` very much looks like a way to store a bool alongside the rest of the structure to reduce the size of the struct by one usize. I'm not against using an Option that way, but if that's the reason, it ought to be written down, and if it isn't the goal, then i'm missing what the goal of this Option is, rather than a `disable_drop: bool`

Reply:
I prefer the `Option`. Setting it to None seems easier to "proofread". I agree that it makes the code less readable however, as the intent is way less clearer than with the boolean solution.

I'll add some comments.

Reply:
you ended up changing it to a bool. If you think an Option is better, you can put it back, just add a 1-2 line comment on the purpose of None

Reply:
Yeah, I thought after packing everything as `drop_without_recycling_permit` it was fine like that.